lib/upload.js

```diff
@@ -821,9 +828,10 @@ class BaseUpload {
       // If the upload length is deferred, the upload size was not specified during
       // upload creation. So, if the file reader is done reading, we know the total
       // upload size and can tell the tus server.
-      if (this.options.uploadLengthDeferred && done) {
+      if (this._uploadLengthDeferred && (!this.options.uploadLengthDeferred || done)) {
```

Comment:
```suggestion
      if (this._uploadLengthDeferred || done) {
```

I think we can just use `this._uploadLengthDeferred` here and don't need to consult the original option anymore.

Reply:
That will change the behaviour quite a bit. As it is suggested right now, there's some problems:
1. We don't always know the size when `this._uploadLengthDeferred` is true since `options.uploadLengthDeferred` could be true. In this case, we'd send wrong info.
2. `done` will make it always pass the header at the end, which I believe isn't how it should work. It should only send the header on the next known request from what I understand.

If we change it to an AND, it'll kinda work, but only send the length header at the end.

The logic in the commit right now expands to:
- `(this._uploadLengthDeferred && !this.options.uploadLengthDeferred)`
 Send header on next opportunity if the request is of known size
- `(this._uploadLengthDeferred && done)`
 And send at the end if it's a stream

Reply:
You're right, my suggestion is wrong. I used `||` instead of `&&` on accident and my suggestion would be:

```suggestion
      if (this._uploadLengthDeferred && done) {
```

> If we change it to an AND, it'll kinda work, but only send the length header at the end.

Yes, that's right and that's how tus-js-client has handled deferred lengths in the past. It always sent the length header once the file source is closed because then we know the file size for sure. I don't see a problem with keeping this behavior. Or do you?

Reply:
Honestly, I'm quite new to the protocol and library so I think you'd know more. But I was looking at this line over at the protocol [documentation](https://tus.io/protocols/resumable-upload#post): 
![image](https://github.com/user-attachments/assets/a51dee04-462c-47fc-b8f3-04c818036606)

That led me to believe that it should be the first possible instance. Although, I guess both ways should work anyway?

Reply:
Hey @Acconut, just want to bump this PR, any thoughts on this? Thanks 😄